File: Android/app/app/src/main/kotlin/edu/bmu/attendance/data/TimetableDateRange.kt
package edu.bmu.attendance.data

import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.TemporalAdjusters
import java.util.Locale

/** Portal date labels for `stu_getBetweenDatesTimetableForStudentSP.json`. */
data class TimetableDateRange(
    val start: LocalDate,
    val end: LocalDate,
) {
    val startLabel: String get() = formatPortalDate(start)
    val endLabel: String get() = formatPortalDate(end)

    companion object {
        private val PORTAL_DATE: DateTimeFormatter =
            DateTimeFormatter.ofPattern("MMM d, yyyy", Locale.US)

        fun formatPortalDate(date: LocalDate): String = date.format(PORTAL_DATE)

        /**
         * Current week Sunday…Saturday, matching the Maitri "Weeks / Current Week" filter.
         * Optional [startLabel]/[endLabel] override either bound when already portal-formatted.
         */
        fun currentWeek(
            startLabel: String? = null,
            endLabel: String? = null,
            today: LocalDate = LocalDate.now(),
        ): TimetableDateRange {
            val sunday = today.with(TemporalAdjusters.previousOrSame(DayOfWeek.SUNDAY))
            val saturday = sunday.plusDays(6)
            val start = startLabel?.let { parsePortalDate(it) } ?: sunday
            val end = endLabel?.let { parsePortalDate(it) } ?: saturday
            return TimetableDateRange(start = start, end = end)
        }

        fun parsePortalDate(label: String): LocalDate =
            LocalDate.parse(label.trim(), PORTAL_DATE)
    }
}

File: IOS/AttendanceCore/Sources/AttendanceCore/TimetableDateRange.swift
import Foundation

/// Portal date labels for `stu_getBetweenDatesTimetableForStudentSP.json`.
/// Matches the gems date filter format (`Aug 2, 2026`), Sunday–Saturday weeks.
public struct TimetableDateRange: Equatable, Sendable {
    public let start: Date
    public let end: Date
    public let startLabel: String
    public let endLabel: String

    public init(start: Date, end: Date) {
        self.start = start
        self.end = end
        self.startLabel = Self.formatPortalDate(start)
        self.endLabel = Self.formatPortalDate(end)
    }

    /// Current week (Sunday…Saturday). Optional overrides replace either bound.
    public static func currentWeek(start: Date? = nil, end: Date? = nil, now: Date = Date()) -> TimetableDateRange {
        let calendar = Calendar(identifier: .gregorian)
        let today = calendar.startOfDay(for: now)
        let weekday = calendar.component(.weekday, from: today) // 1 = Sunday
        let sunday = calendar.date(byAdding: .day, value: -(weekday - 1), to: today) ?? today
        let saturday = calendar.date(byAdding: .day, value: 6, to: sunday) ?? sunday
        return TimetableDateRange(start: start ?? sunday, end: end ?? saturday)
    }

    /// Formats like the Maitri date filter: `Aug 2, 2026` (no zero-padded day).
    public static func formatPortalDate(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone.current
        formatter.dateFormat = "MMM d, yyyy"
        return formatter.string(from: date)
    }
}
